### PR TITLE
Coordination test

### PR DESCRIPTION
**Pull request by carmineceraolo**
Contains the implementation of the coordination test between alternative splicing events in the same gene for all genes.

### CODE REVIEW

File: src/isotools/_utils.py [hunk 1]
```diff
@@ -179,12 +179,13 @@ def _interval_dist(a, b):
     return max([a[0], b[0]])-min([a[1], b[1]])
 
 
-def events_dist(e1, e2):
+def events_dist(e1, e2, sg):
     '''
```
Comment (MatthiasLienhard) [inline src/isotools/_utils.py:183]: as you need the graph, make it a function of the graph class e.g. def events_dist(self, e1, e2)

File: src/isotools/_transcriptome_stats.py [hunk 2]
```diff
@@ -758,5 +761,141 @@ def pairwise_event_test(e1, e2, transcriptome, gene, min_dist=1, test="chi2"):
...
+    :return: a tuple (p_value, stat, Gene, ASE1_type, ASE2_type, ASE1_start, ASE1_end, ASE2_start,
+    ASE2_end, priA_priB, priA_altB, altA_priB, altA_altB), where each element is a list containing the p_values, the statistics, 
+    the gene name, the type of the first ASE, the type of the second ASE, the starting coordinate of the first ASE, 
+    the ending coordinate of the first ASE, the starting coordinate of the second ASE, the ending coordinate of the second ASE, 
+    and the four entries of the contingency table. 
+    '''
+
+    sg = self[gene].segment_graph
+
+    events = list(sg.find_splice_bubbles(types=event_type))
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:800]:
```suggestion
    events = sg.find_splice_bubbles(types=event_type)
```

File: src/isotools/_transcriptome_stats.py [hunk 3]
```diff
@@ -758,5 +761,141 @@ def pairwise_event_test(e1, e2, transcriptome, gene, min_dist=1, test="chi2"):
         p_value = test_res[1]
         test_stat = test_res[0]
 
-        return p_value, test_stat, priA_priB, priA_altB, altA_priB, altA_altB
+        return p_value, test_stat, int(priA_priB-0.01), int(priA_altB-0.01), int(altA_priB-0.01), int(altA_altB-0.01)
+
+
+def gene_coordination_test(self, gene, test="chi2", min_dist=1, min_total=100,
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:767]: This should be a function of the gene

File: src/isotools/_transcriptome_stats.py [hunk 4]
```diff
@@ -758,5 +761,141 @@ def pairwise_event_test(e1, e2, transcriptome, gene, min_dist=1, test="chi2"):
...
+    ASE1_start = []
+    ASE1_end = []
+    ASE2_start = []
+    ASE2_end = []
+    priA_priB = []
+    priA_altB = []
+    altA_priB = []
+    altA_altB = []
+
+    for i, j in list(itertools.combinations(range(len(events)), 2)):
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:816]:
```suggestion
    for i, j in itertools.combinations(range(len(events)), 2):
```
no list needed

File: src/isotools/_transcriptome_stats.py [hunk 5]
```diff
@@ -758,5 +761,141 @@ def pairwise_event_test(e1, e2, transcriptome, gene, min_dist=1, test="chi2"):
...
+
+        test_res = pairwise_event_test(events[i], events[j], self, gene, test=test, min_dist=min_dist)
+
+        if test_res is not None:  # it is none for pairs of events whose distanca is smaller than min_dist
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:820]:
```suggestion
        if sg.event_dist(e1,e2)< min_dist:
            continue
```

File: src/isotools/_transcriptome_stats.py [hunk 6]
```diff
@@ -758,5 +761,141 @@ def pairwise_event_test(e1, e2, transcriptome, gene, min_dist=1, test="chi2"):
...
+    '''
+    Performs pairwise independence test for all pairs of Alternative Splicing Events (ASEs) in a gene.
+
+    For all pairs of ASEs in a gene creates a contingency table and performs an indeppendence test.
+    All ASEs A have two states, pri_A and alt_A, the primary and the alternative state respectivley.
+    Thus, given two events A and B, we have four possible ways in which these events can occur on a transcript, 
+    that is, pri_A and pri_B, pri_A and alt_B, alt_A and pri_B, and alt_A and alt_B. These four values can be put in a contingency table
+    and independence, or coordination, between the two events can be tested.
+
+    ;param gene: Ensemble ID of the gene in which the event happened
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:778]:
```suggestion
    :param gene: Ensemble ID of the gene in which the event happened
```

File: src/isotools/_transcriptome_stats.py [hunk 7]
```diff
@@ -723,24 +723,27 @@ def _filter_event(event, transcriptome, gene, min_total=100, min_alt_fraction=.1
...
     :type test: str
+    :param sg: segment graph
     '''
 
+    sg = kwargs["sg"] if "sg" in kwargs else transcriptome[gene].segment_graph
+
     coverage = transcriptome[gene].coverage.sum(axis=0)
 
-    if events_dist(e1, e2) > min_dist:
+    if events_dist(e1, e2, sg) > min_dist:
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:746]: Do that check before calling the function

File: src/isotools/_transcriptome_stats.py [hunk 8]
```diff
@@ -694,13 +694,13 @@ def _filter_event(event, transcriptome, gene, min_total=100, min_alt_fraction=.1
     '''
     return True if the event satisfies the filter conditions and False otherwise
 
-    :param event: event obtained from .find_splice_bubbles()
-    :param transcriptome: the Transcriptome object on which the events were computed 
+    :param event: Event obtained from .find_splice_bubbles()
+    :param transcriptome: The Transcriptome object on which the events were computed
     ;param gene: Ensemble ID of the gene in which the event happened
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:699]:
```suggestion
    :param gene: Ensemble ID of the gene in which the event happened
```

**Review by MatthiasLienhard — COMMENTED** [review]
Hi Carmine, 
I'd suggest to group the functions to their respective classes, for example the genewise test should be in class Gene.

File: src/isotools/_utils.py (hunk 1 above)
Comment (carmineceraolo) [reply src/isotools/_utils.py]: Done

File: src/isotools/_transcriptome_stats.py (hunk 2 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py:799]: sg.find_splice_bubbles(types=event_type) is a generator and a generator is not subscriptable. If I don't list it, I can't access it in line 817.

Comment (MatthiasLienhard) [reply src/isotools/_transcriptome_stats.py:799]: you make it a list in the next line - no need to do it twice

File: src/isotools/_transcriptome_stats.py (hunk 3 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py]: I made it a function of gene coordination test and change everything that needed to be changed to make it work.

File: src/isotools/_transcriptome_stats.py (hunk 2 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py]: Right. Committed :)

File: src/isotools/_transcriptome_stats.py (hunk 4 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py]: committed.

File: src/isotools/_transcriptome_stats.py (hunk 7 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py]: Ok. Moved the check before the function call.

File: src/isotools/_transcriptome_stats.py (hunk 5 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py]: Done.
    Ok. Moved the check before the function call.

**Comment by carmineceraolo** [comment]
OK, applied all the changes suggested and it seems to work very well.

File: src/isotools/gene.py [hunk 9]
```diff
@@ -415,6 +417,84 @@ def copy(self):
...
+        altA_altB = []
+
+        for i, j in itertools.combinations(range(len(events)), 2):
+
+            if sg.events_dist(events[i], events[j]) < min_dist:
+                continue
+
+            test_res = pairwise_event_test(events[i], events[j], self, test=test)
+
+            if test_res[0] != 1:
```
Comment (MatthiasLienhard) [inline src/isotools/gene.py:478]: We must report also cases with p-value==1, for multiple testing.

File: src/isotools/_transcriptome_stats.py [hunk 10]
```diff
@@ -684,3 +687,103 @@ def rarefaction(self, groups=None, fractions=20, min_coverage=2, tr_filter={}):
...
+    ASE1_start = []
+    ASE1_end = []
+    ASE2_start = []
+    ASE2_end = []
+    priA_priB = []
+    priA_altB = []
+    altA_priB = []
+    altA_altB = []
+
+    for g in self.gene_table().gene_name:
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:765]: why not simply
```suggestion
    for g in self.iter_genes():
```

File: src/isotools/_transcriptome_stats.py [hunk 11]
```diff
@@ -684,3 +687,103 @@ def rarefaction(self, groups=None, fractions=20, min_coverage=2, tr_filter={}):
...
+        test_res = self[g].gene_coordination_test(test=test, min_dist=min_dist, min_total=min_total,
```
Comment (MatthiasLienhard) [inline src/isotools/_transcriptome_stats.py:766]: gene_coordination_test should return a list of test result tuples, one tuple per comparison. This makes it much simpler
```suggestion
        test_res.extend(self[g].gene_coordination_test(test=test, min_dist=min_dist, min_total=min_total,
```

File: src/isotools/gene.py [hunk 12]
```diff
@@ -415,6 +417,84 @@ def copy(self):
...
+        priA_altB = []
+        altA_priB = []
+        altA_altB = []
+
+        for i, j in itertools.combinations(range(len(events)), 2):
+
+            if sg.events_dist(events[i], events[j]) < min_dist:
+                continue
+
+            test_res = pairwise_event_test(events[i], events[j], self, test=test)
```
Comment (MatthiasLienhard) [inline src/isotools/gene.py:476]:
```suggestion
            test_res.append(pairwise_event_test(events[i], events[j], self, test=test))
```

File: src/isotools/gene.py [hunk 13]
```diff
@@ -415,6 +417,84 @@ def copy(self):
...
+        ASE1_type, ASE2_type = [], []
+        ASE1_start = []
+        ASE1_end = []
+        ASE2_start = []
+        ASE2_end = []
+        priA_priB = []
+        priA_altB = []
+        altA_priB = []
+        altA_altB = []
+
```
Comment (MatthiasLienhard) [inline src/isotools/gene.py:470]: test_res=[]

File: src/isotools/gene.py [hunk 14]
```diff
@@ -415,6 +417,84 @@ def copy(self):
...
+                ASE2_type.append(events[j][4])
+                ASE1_start.append(sg[events[i][2]].start)  # starting coordinate of event 1
+                ASE1_end.append(sg[events[i][3]].end)  # ending coordinate of event 1
+                ASE2_start.append(sg[events[j][2]].start)  # starting coordinate of event 2
+                ASE2_end.append(sg[events[j][3]].end)  # ending coordinate of event 2
+                priA_priB.append(test_res[2])
+                priA_altB.append(test_res[3])
+                altA_priB.append(test_res[4])
+                altA_altB.append(test_res[5])
+
```
Comment (MatthiasLienhard) [inline src/isotools/gene.py:492]:
```suggestion
return test_res
```

File: tests/coordination_test.py [hunk 15]
```diff
@@ -0,0 +1,8 @@
+
+def test_coordination_T(example_gene_coor,res_coor):
+    res=example_gene_coor.gene_coordination_test(test="fisher")
+    assert all([res[i]==res_coor[i] for i in range(2,13)])
```
Comment (MatthiasLienhard) [inline tests/coordination_test.py:4]: maybe also test p-value >0 and pvalue <.05

File: src/isotools/gene.py [hunk 16]
```diff
@@ -415,6 +417,84 @@ def copy(self):
...
+        Alternative Splicing Events for the pair to be tested
+        :type min_dist: int
+        :param min_total: The minimum total number of reads for an event to pass the filter
+        :type min_total: int
+        :param min_alt_fraction: The minimum fraction of read supporting the alternative
+        :type min_alt_frction: float
+        :param event_type:  A tuple with event types to test. Valid types are
+        (‘ES’,’3AS’, ‘5AS’,’IR’ or ‘ME’, ‘TSS’, ‘PAS’). Default is ("ES", "5AS", "3AS", "IR", "ME")
+
+        :return: a tuple (p_value, stat, Gene, ASE1_type, ASE2_type, ASE1_start, ASE1_end, ASE2_start,
```
Comment (MatthiasLienhard) [inline src/isotools/gene.py:444]: Try doing it the other way round (a list of tuples) - this makes it much easier

File: src/isotools/_transcriptome_stats.py [hunk 17]
```diff
@@ -684,3 +687,103 @@ def rarefaction(self, groups=None, fractions=20, min_coverage=2, tr_filter={}):
...
+    C = pd.DataFrame({"A_pri": (0, 0), "A_alt": (0, 0)})
+
+    C.index = ("B_pri", "B_alt")
+
+    for m, n in itertools.product(range(2), range(2)):
+        tr_IDs = list(set(e1[m]) & set(e2[n]))
+        c_mn = sum([coverage[ID] for ID in tr_IDs])+0.01
+        C.iloc[n, m] = c_mn
+
+    TEST = TESTS[test]
```
[the remark this answers is not in the document]
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py:718]: Fixed

File: src/isotools/_transcriptome_stats.py [hunk 18]
```diff
@@ -684,3 +687,103 @@ def rarefaction(self, groups=None, fractions=20, min_coverage=2, tr_filter={}):
...
+    (the chi squared statistic if the chi squared test is used and the odds-ratio if the Fisher
+    test is used), the gene name, the type of the first ASE, the type of the second ASE, the
+    starting coordinate of the first ASE, the ending coordinate of the first ASE, the starting
+    coordinate of the second ASE, the ending coordinate of the second ASE,
+    and the four entries of the contingency table.
+    '''
+    p_value = []
+    stat = []
+    Gene = []
+    ASE1_type, ASE2_type = [], []
```
[the remark this answers is not in the document]
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py:755]: Fixed

File: src/isotools/gene.py (hunk 9 above)
Comment (carmineceraolo) [reply src/isotools/gene.py:478]: Fixed

File: src/isotools/_transcriptome_stats.py (hunk 10 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py:765]: Fixed.

File: src/isotools/_transcriptome_stats.py (hunk 11 above)
Comment (carmineceraolo) [reply src/isotools/_transcriptome_stats.py:766]: Resolved. But then I still do 13 list comprehensions to later use in the construction of the pandas df. If frugality was your intention, I might have to look at different ways to construct a pandas df before merging. If better readability was what you were aiming for, I think it now still looks better than before.

File: src/isotools/gene.py (hunk 12 above)
Comment (carmineceraolo) [reply src/isotools/gene.py:476]: Fixed

File: src/isotools/gene.py (hunk 14 above)
Comment (carmineceraolo) [reply src/isotools/gene.py:492]: Fixed.

File: src/isotools/gene.py (hunk 13 above)
Comment (carmineceraolo) [reply src/isotools/gene.py:470]: Fixed.

File: tests/coordination_test.py (hunk 15 above)
Comment (carmineceraolo) [reply tests/coordination_test.py:4]: Added this as well.

**Pull request merged**